crates/turborepo-cache/src/cache_archive/create.rs

```diff
@@ -0,0 +1,174 @@
+use std::{
+    backtrace::Backtrace,
+    fs,
+    fs::{File, OpenOptions},
+    io::{BufWriter, Read},
+    path::Path,
+};
+
+use tar::{EntryType, Header};
+use turbopath::{AbsoluteSystemPath, AnchoredSystemPath, RelativeUnixPathBuf};
+
+use crate::CacheError;
+
+// We use an enum to get around Rust's generic restrictions
+// i.e. you can't have a function that can return two different
+// versions of a generic type like Vec<u32> and Vec<u64>
+enum CacheArchive<'a> {
+    Compressed(tar::Builder<zstd::Encoder<'a, BufWriter<File>>>),
+    Uncompressed(tar::Builder<BufWriter<File>>),
+}
+
+impl<'a> CacheArchive<'a> {
+    // Appends data to tar builder.
+    fn append_data(
+        &mut self,
+        header: &mut Header,
+        path: impl AsRef<Path>,
+        body: impl Read,
+    ) -> Result<(), CacheError> {
+        match self {
+            CacheArchive::Compressed(builder) => Ok(builder.append_data(header, path, body)?),
+            CacheArchive::Uncompressed(builder) => Ok(builder.append_data(header, path, body)?),
+        }
+    }
+
+    // Makes a new CacheArchive at the specified path
+    // Wires up the chain of writers:
+    // tar::Builder -> zstd::Encoder (optional) -> BufWriter -> File
+    fn create(path: &AbsoluteSystemPath) -> Result<Self, CacheError> {
+        let mut options = OpenOptions::new();
+
+        let file = options
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(path.as_path())?;
+
+        // Flush to disk in 1mb chunks.
+        let file_buffer = BufWriter::with_capacity(2usize.pow(20), file);
+
+        let is_compressed = path.as_path().ends_with(".zst");
```

Comment:
Instead of `as_path`, let's add support for `.ends_with` to `AbsoluteSystemPath`.

As a general comment, we should aim for not importing `fs` and `path` and friends. It may not always be possible, but that's the ideal.